### ISSUE TITLE
vendor in different modules cause inconsistencies

### ISSUE THREAD

**Issue by jeffisabel** [golang/go#26583]
Please answer these questions before submitting your issue. Thanks!

### What version of Go are you using (`go version`)?
go version go1.10.3 linux/amd64

### Does this issue reproduce with the latest release?
Yes.

### What operating system and processor architecture are you using (`go env`)?
GOARCH="amd64"
GOHOSTARCH="amd64"
GOHOSTOS="linux"

### What did you do?
The source tree looks like  (`tree`  from GOPATH)
```
.
`-- src
    |-- a
    |   |-- a.go
    |   `-- vendor
    |       `-- c
    |           `-- c.go
    `-- b
        |-- b.go
        `-- vendor
            `-- c
                `-- c.go
```

Module C (c.go) is a library can do something:
```
package c

import "errors"

var (
	ErrorX = errors.New("ErrorX")
	ErrorY = errors.New("ErrorX")
)

func DoSomething() error {
	// do something
	return ErrorX
}

func IsErrorX(err error) bool {
	return err == ErrorX
}
```
Module B (b.go) is another library which is essentially a wrapper of module C

```
package b

import (
	"c"
)

func DoSomethingElse() error {
	// do something else
	return c.DoSomething()
}
```

Module A (a.go) uses both module B and module C and checks error returned:
```
package main

import (
	"b"
	"c"
	"fmt"
)

func main() {
	err := b.DoSomethingElse()
	fmt.Printf("%v\n", c.IsErrorX(err))
}
```

Please see below for explanation;

### What did you expect to see?
```
$ ./a
true
```

In above code snippet, actually C's DoSomethin() returned an "ErrorX", and B's DoSomethingElse() just pass this error directly to A, so IsErrorX() is supposed to return `true`. However, we found that the error returned is not the same instance as the static one declared in C.

In other words, go compiler created two module C instances, although both instances are identical. In fact they have same string value ("ErrorX"), but they are different instance . It seems to me that determining the error type by comparing with a static pointer is fast by dangerous. I encountered this issue firstly with GORM (https://github.com/jinzhu/gorm). Then I noticed that this error.New() style is used in several standard libraries (e.g. src/os/error) and it is likely where this style comes from.

Still, if this is not a bug, is there any workaround? (e.g. some hints on compile/link flags or code patterns suggestion would be appreciated)

Thank you!

### What did you see instead?
```
$ go build a/...
$ ./a
false
```

**Comment by bradfitz** [comment]
That's expected. Those are two independent packages and values from errors.New are always unique (https://play.golang.org/p/1h1otlmvQSm).

If you want this to work, you need to move the vendored package up so both `a` and `b` use the same `c`.

Try out Go 1.11beta2's new module support, BTW. The new tooling might make dealing with all this easier.

**Comment by jeffisabel** [comment]
Thanks for answering.

I understand that moving Module C up does solve the problem. But the situation in real life is that Module B might be a 3rd party library and we are not supposed to modify its vendor. And there might be a bunch of Module B(s) in project. What makes situation worse is that the problem might hide deep in vendor's vendor's vendor and we don't even know it is there.

Still, thanks for pointer to Go module. I'll try it out.
